### PR TITLE
Python-native typing support

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
Please see the proposal doc at https://docs.google.com/document/d/17rNKg6Uvow8CrECaPff96Tarr87P2fn4ilf_Tv2lYd4
https://flytecookbook.readthedocs.io/en/latest/

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

### CODE REVIEW

File: flytekit/annotated/stuff.py [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+
+
+def task(
+        _task_function=None,
+        outputs: List[str]=None,
+        cache_version='',
+        retries=0,
+        interruptible=None,
+        deprecated='',
+        storage_request=None,
```
Comment (kumare3) [inline flytekit/annotated/stuff.py:305]: i want to club the requests and limits as structs. This simplifies adding new ones

File: flytekit/annotated/stuff.py [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+        cpu_request=None,
+        gpu_request=None,
+        memory_request=None,
+        storage_limit=None,
+        cpu_limit=None,
+        gpu_limit=None,
+        memory_limit=None,
+        cache=False,
```
Comment (kumare3) [inline flytekit/annotated/stuff.py:313]: samething for cache?

File: flytekit/annotated/sample.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+@task(outputs=['s_out'])
+def x(s: int) -> int:
+    return s + 1
+
+
+@workflow(outputs=["real_b"])
+def my_workflow() -> WorkflowOutputs:
+    # a = x(s=3)
+    b = x(s=x(s=3))
```
Comment (kumare3) [inline flytekit/annotated/sample.py:12]: can we get a little more complex example, where we have 3 nodes, multiple outputs and inputs?

**Review by kumare3 — APPROVED** [review]
Oh My God! Hurrah!

**Review by katrogan — APPROVED** [review]

**Pull request merged**